pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -114,6 +123,61 @@ func podKeyGet(pod *corev1.Pod) string {
 	return pod.Namespace + "/" + pod.Name
 }
 
+// buildVFDeviceIDListPerPod is a helper function to build a cache structure with the
+// list of all the PCI addresses allocated per Pod based on their resource requests (in Pod spec).
+// When there is a request for a VF resource (to associate it for a secondary network interface),
+// getUnusedSriovVFDeviceIDPerPod will use this cache information to pick up a unique PCI address
+// which is still not associated with a network device name.
+// NOTE: buildVFDeviceIDListPerPod is called only if a Pod specific VF to Interface mapping cache
+// was not build earlier. Sample initial entry per Pod: "{18:01.1,""},{18:01.2,""},{18:01.3,""}"
+func (pc *PodController) buildVFDeviceIDListPerPod(podName, podNamespace string) error {
+	key := podNamespace + "/" + podName
+	pc.vfDeviceIDCacheMutex.RLock()
+	if pc.vfDeviceIDUsageMap[key] != nil {
+		pc.vfDeviceIDCacheMutex.RUnlock()
+		return nil
+	}
+	// When a pod gets created for the first time, vfDeviceIDUsageMap will be empty. So we need to
+	// unlock the RLock here. On a repeated/secondary call to buildVFDeviceIDListPerPod,
+	// vfDeviceIDUsageMap expected to be non nil (gets unlocked just before the above return).
+	// Can't use defer, as we look to acquire a write lock below.
+	pc.vfDeviceIDCacheMutex.RUnlock()
```

Comment:
Maybe sync.Map is simpler to avoid multiple lock/unlock. No strong preference.

Reply:
that is the better fit for this need @jianjuns . reworked to use sync.Map and tested the fix again. it works good as expected. 
all the above comments were addressed as well.

Reply:
2022-09-22T19:02:58Z [debug] Elected as leader, do processing
2022-09-22T19:02:58Z [debug] IPManagement -- mode: 0 / containerID: 3ce2eb745c5d77640de159d22d72af65d1fd260987e0e084209158c13566daf8 / podRef: default/testpodsec120
2022-09-22T19:02:58Z [error] IPAM error reading pool allocations (attempt: 0): k8s pool initialized
2022-09-22T19:02:58Z [debug] IterateForAssignment input >> ip: 148.14.25.0 | ipnet: {148.14.25.0 ffffff00} | first IP: 148.14.25.1 | last IP: 148.14.25.254
2022-09-22T19:02:58Z [debug] Reserving IP: |148.14.25.1 3ce2eb745c5d77640de159d22d72af65d1fd260987e0e084209158c13566daf8|
2022-09-22T19:02:58Z [debug] OverlappingRangewide allocation check for IP: 148.14.25.1
2022-09-22T19:02:58Z [debug] K8s UpdateOverlappingRangeAllocation success on allocate: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:148.14.25.1 GenerateName: Namespace:kube-system SelfLink: UID:7d3b00db-c2bd-41f2-82f5-1dd57c0365b3 ResourceVersion:102476 Generation:1 CreationTimestamp:2022-09-22 19:02:58 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:whereabouts Operation:Update APIVersion:whereabouts.cni.cncf.io/v1alpha1 Time:2022-09-22 19:02:58 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:spec":{".":{},"f:containerid":{},"f:podref":{}}} Subresource:}]} Spec:{ContainerID:3ce2eb745c5d77640de159d22d72af65d1fd260987e0e084209158c13566daf8 PodRef:default/testpodsec120}}
2022-09-22T19:02:58Z [debug] OnStoppedLeading() called
2022-09-22T19:02:58Z [debug] Finished leader election
2022-09-22T19:02:58Z [debug] IPManagement: {148.14.25.1 ffffff00}, <nil>
I0922 19:02:58.280010       1 controller.go:333] "INTEL - UNUSED VF ID:" VF ID="0000:18:04.4"
I0922 19:02:58.348523       1 sriov.go:113] Configured interfaces for container 3ce2eb745c5d77640de159d22d72af65d1fd260987e0e084209158c13566daf8; hostIface: {Name:ens785f0v28 Mac:1a:a6:69:d4:13:b0 Sandbox:}, containerIface: {Name:eth2 Mac:1a:a6:69:d4:13:b0 Sandbox:/host/proc/1130811/ns/net}
I0922 19:04:37.225512       1 server.go:539] Received CmdDel request cni_args:{container_id:"60d84cbdc5d847d156cd2f16ae647793225425a130e39b777d89b54d843f68db"  netns:"/proc/1110870/ns/net"  ifname:"eth0"  args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=testpodsec-intel;K8S_POD_INFRA_CONTAINER_ID=60d84cbdc5d847d156cd2f16ae647793225425a130e39b777d89b54d843f68db"  path:"/opt/cni/bin"  network_configuration:"{\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}"}
I0922 19:04:37.231339       1 server.go:563] Deleted IP addresses for container 60d84cbdc5d847d156cd2f16ae647793225425a130e39b777d89b54d843f68db
I0922 19:04:37.233149       1 pod_configuration.go:537] Removed interfaces for container 60d84cbdc5d847d156cd2f16ae647793225425a130e39b777d89b54d843f68db
I0922 19:04:37.339714       1 server.go:569] CmdDel for container 60d84cbdc5d847d156cd2f16ae647793225425a130e39b777d89b54d843f68db succeeded
I0922 19:05:00.860464       1 server.go:425] Received CmdAdd request cni_args:{container_id:"2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d"  netns:"/proc/1132614/ns/net"  ifname:"eth0"  args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=testpodsec-intel;K8S_POD_INFRA_CONTAINER_ID=2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d"  path:"/opt/cni/bin"  network_configuration:"{\"cniVersion\":\"0.3.0\",\"ipam\":{\"type\":\"host-local\"},\"name\":\"antrea\",\"type\":\"antrea\"}"}
I0922 19:05:00.866924       1 server.go:495] "Allocated IP addresses" container="2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d" result=&{Result:{CNIVersion:0.4.0 Interfaces:[] IPs:[{Version:4 Interface:<nil> Address:{IP:192.168.0.8 Mask:ffffff00} Gateway:192.168.0.1}] Routes:[] DNS:{Nameservers:[] Domain: Search:[] Options:[]}} VLANID:0}
I0922 19:05:00.939502       1 pod_configuration.go:265] Configured interfaces for container 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d
I0922 19:05:00.939540       1 server.go:522] CmdAdd for container 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d succeeded
I0922 19:05:01.893044       1 controller.go:395] "Secondary Network attached to Pod" network=&{Name:virtual-net2 Namespace:default IPRequest:[] MacRequest: InfinibandGUIDRequest: InterfaceRequest:eth1 PortMappingsRequest:[] BandwidthRequest:<nil> CNIArgs:<nil> GatewayRequest:[]} Pod="default/testpodsec-intel"
2022-09-22T19:05:01Z [debug] ADD - IPAM configuration successfully read: {Name:virtual-net2 Type:whereabouts Routes:[] Datastore:kubernetes Addresses:[] OmitRanges:[] DNS:{Nameservers:[] Domain: Search:[] Options:[]} Range:148.14.25.0/24 RangeStart:148.14.25.0 RangeEnd:<nil> GatewayStr: EtcdHost: EtcdUsername: EtcdPassword:********* EtcdKeyFile: EtcdCertFile: EtcdCACertFile: LeaderLeaseDuration:1500 LeaderRenewDeadline:1000 LeaderRetryPeriod:500 LogFile: LogLevel: OverlappingRanges:true SleepForRace:0 Gateway:<nil> Kubernetes:{KubeConfigPath:/host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig K8sAPIRoot:} ConfigurationPath: PodName:testpodsec-intel PodNamespace:default}
2022-09-22T19:05:01Z [debug] Beginning IPAM for ContainerID: 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d
2022-09-22T19:05:01Z [debug] Started leader election
I0922 19:05:01.974569     541 leaderelection.go:248] attempting to acquire leader lease kube-system/whereabouts...
I0922 19:05:01.992267     541 leaderelection.go:258] successfully acquired lease kube-system/whereabouts
2022-09-22T19:05:01Z [debug] OnStartedLeading() called
2022-09-22T19:05:01Z [debug] Elected as leader, do processing
2022-09-22T19:05:01Z [debug] IPManagement -- mode: 0 / containerID: 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d / podRef: default/testpodsec-intel
2022-09-22T19:05:02Z [debug] IterateForAssignment input >> ip: 148.14.25.0 | ipnet: {148.14.25.0 ffffff00} | first IP: 148.14.25.1 | last IP: 148.14.25.254
2022-09-22T19:05:02Z [debug] Reserving IP: |148.14.25.2 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d|
2022-09-22T19:05:02Z [debug] OverlappingRangewide allocation check for IP: 148.14.25.2
2022-09-22T19:05:02Z [debug] K8s UpdateOverlappingRangeAllocation success on allocate: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:148.14.25.2 GenerateName: Namespace:kube-system SelfLink: UID:fca1a47f-c1d4-4955-9405-a1ce82eaaf4c ResourceVersion:102666 Generation:1 CreationTimestamp:2022-09-22 19:05:02 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[{Manager:whereabouts Operation:Update APIVersion:whereabouts.cni.cncf.io/v1alpha1 Time:2022-09-22 19:05:02 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:spec":{".":{},"f:containerid":{},"f:podref":{}}} Subresource:}]} Spec:{ContainerID:2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d PodRef:default/testpodsec-intel}}
2022-09-22T19:05:02Z [debug] OnStoppedLeading() called
2022-09-22T19:05:02Z [debug] Finished leader election
2022-09-22T19:05:02Z [debug] IPManagement: {148.14.25.2 ffffff00}, <nil>
I0922 19:05:02.032616       1 controller.go:333] "INTEL - UNUSED VF ID:" VF ID="0000:18:02.2"
I0922 19:05:02.104640       1 sriov.go:113] Configured interfaces for container 2ad8370d64106173669e2d54e4571fffa9f9170dc7a24f97f69f1cd9ed2b6e7d; hostIface: {Name:ens785f0v10 Mac:96:62:1a:7a:51:0f Sandbox:}, containerIface: {Name:eth1 Mac:96:62:1a:7a:51:0f Sandbox:/host/proc/1132614/ns/net}
root@vcesd2:/home/vce/August29_2022/antrea#
[8:27](https://kubernetes.slack.com/archives/C02C3DJQVJT/p1663903649298699)
works good. Added that INTEL log only for my testing needs. removed it on the PR pushed earlier.
[8:28](https://kubernetes.slack.com/archives/C02C3DJQVJT/p1663903726842109)
kubectl exec -it testpodsec-intel -- ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
3: eth0@if96: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1450 qdisc noqueue
    link/ether 4e:d8:ce:17:f1:09 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.8/24 brd 192.168.0.255 scope global eth0
       valid_lft forever preferred_lft forever
66: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc mq qlen 1000
    link/ether 96:62:1a:7a:51:0f brd ff:ff:ff:ff:ff:ff
    inet 148.14.25.2/24 brd 148.14.25.255 scope global eth1
       valid_lft forever preferred_lft forever